torch/csrc/jit/passes/device_type_analysis.cpp

```diff
@@ -0,0 +1,211 @@
+#include <ATen/core/interned_strings.h>
+#include <ATen/core/jit_type.h>
+#include <c10/core/Device.h>
+#include <c10/util/ArrayRef.h>
+#include <c10/util/Optional.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/device_type_analysis.h>
+#include <torch/library.h>
+#include <memory>
+#include <utility>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+using Tensor = at::Tensor;
+using Device = at::Device;
+
+using PropRule = std::function<bool(Node*)>;
+/*
+A Propagation Rule takes the Node, and
+applies the relevant properties to the Tensor outputs
+of the Node (based on the rule itself)
+
+Returns: Bool indicating if anything was changed
+*/
+
+bool setDeviceType(Value* value, c10::optional<Device> device) {
+  auto tensor_type = value->type()->cast<TensorType>();
+  TORCH_INTERNAL_ASSERT(tensor_type, "Expecting a tensor type");
+  bool changed = tensor_type->device() != device;
+  if (changed) {
+    value->setType(tensor_type->withDevice(device));
+  }
+  return changed;
+}
+
+bool setReturnsToDevice(Node* n, c10::optional<Device> device) {
+  bool changed = false;
+  for (Value* out : n->outputs()) {
+    auto tensor_type = out->type()->cast<TensorType>();
+    if (!tensor_type) {
+      continue;
+    }
+    changed |= setDeviceType(out, device);
+  }
+  return changed;
+}
+
+bool propWithNoDevice(Node* n) {
+  // Figure out what the common device to propagate is
+  // Types of tensors must match, except CPU zerodim, which any
+  // other type can overwrite
+
+  c10::optional<Device> device;
+  bool seen_any_device = false;
+  bool only_seen_cpu_zerodim = false;
+
+  for (Value* inp : n->inputs()) {
+    auto tensor_type = inp->type()->cast<TensorType>();
+    if (!tensor_type) {
+      continue;
+    }
+
+    bool is_zerodim = tensor_type->symbolic_sizes().rank().value_or(-1) == 0;
```

Comment:
`is_zerodim` - don't we want to conservatively check if it's zerodim? Here we are saying, if we don't know the rank, assume it is not zerodim

Reply:
Yes, the conservative case is that it is not zerodim, as non-zerodim tensors must have the same device, whereas zerodim CPU tensors can be used with other tensors.

Reply:
We're not being conservative here - we're saying that if we don't know the rank, we assume `is_zerodim = -1 ==0` which is false. The conservative assumption is that it is zerodim

Reply:
We're not being conservative here - 
```
@torch.jit.script
def foo(x, y):
   return x + y
```

if we set x to be cpu, but not have a rank, then we will have `is_zerodim = (-1 == 0) = False`. Then, we'll propagate the output type to be cpu because it's not zerodim. 
`foo(torch.tensor(0.3), torch.rand([2, 2]).cuda()) -> cuda tensor`. the analysis is not being conservative here.

Reply:
Incorrect. The current code would return `c10::nullopt` as the device type in the case without shape info because there are two conflicting types.

Reply:
If you want more extensive testing than the below link for zerodim, let me know what you want so that I can assure you that the code is correct.

https://github.com/pytorch/pytorch/pull/69052/files#diff-8c9c2b2ce38114d5a65e06e990ae810ea68f4dd2f54e80fc65ec7cf3d53b5c2eR105-R147

Reply:
> Incorrect. The current code would return c10::nullopt as the device type in the case without shape info because there are two conflicting types.

In this case we are leaving `y` not set. So we're inferring `cpu`, even though as the example shows there is a valid invocation that would produce a `cuda` tensor.